mp/src/materialsystem/stdshaders/pbr_dx9.cpp

```diff
@@ -0,0 +1,524 @@
+        if (info.Normal != -1 && params[info.Normal]->IsDefined())
+        {
+            LoadBumpMap(info.Normal); // normal is a bump map
+        }
+
+        // setting up mrao map
+        if (info.Mrao != -1 && params[info.Mrao]->IsDefined())
+        {
+            LoadTexture(info.Mrao);
+        }
+
+        // setting up emissive texture
+        if (info.Emissive != -1 && params[info.Emissive]->IsDefined())
+        {
+            LoadTexture(info.Emissive, TEXTUREFLAGS_SRGB); // emissive is srgb
+        }
+
+        // setting up environment map
+        if (info.Envmap != -1 && params[info.Envmap]->IsDefined())
+        {
+            // theoretically this should allow the game to access lower mips, but it doesn't seem to do anything
+            int flags = TEXTUREFLAGS_ALL_MIPS | TEXTUREFLAGS_NOLOD;
+            if (g_pHardwareConfig->GetHDRType() == HDR_TYPE_NONE)
+            {
+                flags |= TEXTUREFLAGS_SRGB; // envmap is only srgb with hdr disabled?
+            }
+
+            // if the hardware doesn't support cubemaps, use spheremaps instead
+            if (!g_pHardwareConfig->SupportsCubeMaps())
+            {
+                SET_FLAGS(MATERIAL_VAR_ENVMAPSPHERE);
+            }
+
+            // done like this so the user could set $envmapsphere themselves if they wanted to
+            if (!IS_FLAG_SET(MATERIAL_VAR_ENVMAPSPHERE))
+            {
+                LoadCubeMap(info.Envmap, flags);
+            }
+            else
+            {
+                LoadTexture(info.Envmap, flags);
+            }
+        }
+
+        // setting up flashlight texture
+        if (info.FlashlightTexture != -1 && params[info.FlashlightTexture]->IsDefined())
+        {
+            LoadTexture(info.FlashlightTexture, TEXTUREFLAGS_SRGB); // flashlight texture is srgb
+        }
+    };
+
+    // drawing the shader
+    SHADER_DRAW
+    {
+        PBR_DX9_Vars_t info;
+        SetupVars(info);
+
+        // setting up some booleans
+        bool bIsAlphaTested     = IS_FLAG_SET(MATERIAL_VAR_ALPHATEST) != 0;
+        bool bHasAlbedo         = (info.Albedo != -1) && params[info.Albedo]->IsTexture();
+        bool bHasNormal         = (info.Normal != -1) && params[info.Normal]->IsTexture();
+        bool bHasMrao           = (info.Mrao != -1) && params[info.Mrao]->IsTexture();
+        bool bHasEmissive       = (info.Emissive != -1) && params[info.Emissive]->IsTexture();
+        bool bHasEnvmap         = (info.Envmap != -1) && params[info.Envmap]->IsTexture();
+        bool bHasFlashlight     = UsingFlashlight(params);
+        bool bLightmapped       = IS_FLAG_SET(MATERIAL_VAR_MODEL) == 0;
+
+        // determining whether we're dealing with a fully opaque material
+        BlendType_t nBlendType = this->EvaluateBlendRequirements(info.Albedo, true);
+        bool bFullyOpaque = (nBlendType != BT_BLENDADD) && (nBlendType != BT_BLEND) && !bIsAlphaTested;
+
+        if (this->IsSnapshotting()) // shadow state
```

Comment:
I'd prefer if you used the `SHADOW_STATE`/`DYNAMIC_STATE` macros

Reply:
sure, I'll do that